docs: add auto scaler proposal for inference workloads (#1172)
kaito-project/kaito: website/docs/proposals/20250620-auto-scaler-for-inference-workloads.md

---
title: Auto Scaler For Inference Workloads In Kaito
authors:
  - "@rambohe-ch"
reviewers:
  - "@Fei-Guo"
  - "@helayoty"
  - "@zhuangqh"
creation-date: 2025-06-10
last-updated: 2025-06-20
status: provisional
see-also:
---

# Title

Auto Scaler for inference workloads in Kaito

## Summary

As the number of waiting inference requests increase, It is necessary to scale more inference instances in order to preventing to block inference requests. on the other hand, If the number of waiting inference requests declines, we should consider to reduce inference instances for improving gpu resource utilization.
Native Kubernetes has provided HPA capability to scale workload instance automatically as the metrics change. but HPA depends the third-party components(like prometheus, prometheus-adapter, etc.) to collect custom metrics from the source pods.

> In this proposal, we hope to support a customized auto-sacler which is specialized for scaling GPU worklods for kaito. The auto-scaler is designed for a minimalistic configuration experience, with most parameters pre-tuned for optimal performance. This allows users to easily get started without requiring specialized knowledge of LLM.

Author: we want to provide a minimal auto-scaler to user in the first stage, and enrich the auto-scaler features in future versions. at the same time, more configurations(lile cooldown, stablization window, scale step, etc.) maybe exposed to user.

## Proposal

### Auto-Scaler Architecture

The llm-auto-scaler component scrapes metrics from inference pod according to configurations in LLMAutoScaler CRD, and scaler controller calculate desired replicas by integrating scraped metrics and scale strategy,
then scale workspace replicas through /scale subresource API. The detailed auto-scaler architecture is shown in the following figure:

![auto-scaler](../../static/img/llm-auto-scaler.png)

- LLMAutoScaler CRD: is used as auto-scaler configuration(including scale strategy, target reference, etc.) for specified workspace resource.
- Metrics Scraper: a module in auto-scaler of llm-auto-scaler and used for scraping metrics from inference pod.
- Scale Strategy: is used to specify scaler, like basic scaler, cron scaler. different scaler have different algorithm to calculate desired replicas. In this proposal, only basic scaler will be supported. and more strategies will be supported in future versions.
- Scaler Controller: is used for integrating metrics scraper and scaler strategy, also including invoke scale subresource API of workspace.

> ### LLMAutoScaler CRD

Reviewer: Could we rename it to `Autoscaler` because KAITO could be serving other GenAI models other than LLM in the future 😄

Author: maybe we should add a prefix for AutoScaler, and we can discuss the name tomorrow and find a more suitable one.

Reviewer 2: How about GMA (pronounced like Jemma) [GenAI Model Autoscaler]? Quite a bit of acronym inception here, but this name could be more inclusive of a wider range of model types beyond LLMs.

```
type ProtocolType string

const (
	HTTP  ProtocolType = "http"
	HTTPS ProtocolType = "https"
)

// MetricIdentifier defines the way to fetch the specific metric
type MetricIdentifier struct {
	// Name identifies the specific metric to monitor.
	// If unset, vllm:num_requests_waiting will be used.
	Name string

	// selector is the string-encoded form of a standard kubernetes label selector for the given metric.
	// if unset, a selector related to ScaleTargetRef will be configured(like workspace specified labels).

> 	Selector *metav1.LabelSelector

Reviewer: I don't think we need this if we implement [labelSelectorPath](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#scale-subresource) from scale subresource in KAITO's Workspace

Author: I'wll remove it.

	// Protocol specify the protocol for accessing pods, http and https are supported.
	// if unset, http will be used.
	Protocol ProtocolType

	// Port specify the port of pods /metrics endpoint.
	// if unset, 5000 will be used.
	Port string
}

type MetricThreshold struct {

> 	// High means the uuper threshold, when the value of the monitored metric exceeds this number,

Author: fixed

	// the autoscaler will decide to scale up.
	High int32

	// Low mens the lower threshold. when the value of the monitored metric drops below this number,
	// the autoscaler will scale down.

> 	Low int32

Reviewer: Curious why we opt for high-low strategy rather than [HPA's formula](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#algorithm-details). Is it about tolerating a wider range of metric fluctuation?

![image](https://github.com/user-attachments/assets/b5e60db7-5537-4ab8-9241-3fd75d955d1b)

Author: There are two reasons:
1. metric `vllm:num_requests_waiting` is used for calculating scaling action. this is a count value, so thershold value is more suitable than percentage like HPA toleration.
2. In native HPA, user need to define a target value and two toleration values. and only two threshold value should be specified in our solution. compared to the native HPA definition, threshold definition is more clear.

Reviewer 3: I think LLM application is different from traditional app. It typically cost around 5min to scale an A100 machine.
We need to be more careful when scaling down. An additional, more conservative low threshold can avoid meaningless scaling back and forth.

- LowThreshold: expected low threshold of waiting requests, related field: `LLMAutoScaler.Spec.Metrics[0].Threshold.Low`
- ScaleUpStep: the scale step of scaling up action, default value is 1
- ScaleDownStep: the scale step of scaling down action, default value is 1
- UpStabilizationWindowSeconds: the stabilization window seconds of scaling up action, default value is 0
- DownStabilizationWindowSeconds: the stabilization window seconds of scaling down action, default value is 300
- UpCoolDownSeconds: the cool down seconds of scaling up action, default value is 600
- DownCoolDownSeconds: the cool down seconds of scaling down action, default value is 1800
- LastScaleTime: the timestamp for the last scaling action, related field: `LLMAutoScaler.Status.LastScaleTime`

Outputs:

- DesiredReplicas: Desired number of replicas for target workload. and the value will be used for scaling workload through /Scale subresource api.

```
// 1. calculate the elapsed time for cooldown check
coolddownElapsed := now.Sub(LastScaleTime)

// 2. scale up logic
if CurrentWaitingRequests > HighThreshold {
	// check stablization window
	if UpStabilizationWindowSeconds > 0 {

> 		windowMetrics := filterMetricsWithinWindow(queueHistory, UpStabilizationWindowSeconds)

Author: I think this is a implmentation issue. maybe we can discuss the details in the code review.

> 		windowMetrics := filterMetricsWithinWindow(queueHistory, UpStabilizationWindowSeconds)

Author: @chewong @ritazh I've updated pseudocode related to queueHistory.

### Metrics Scraper

- Metrics Scraper fetches specified metrics from pods' /metrics endpoint according to `LLMAutoScaler.Spec.Metrics` at 15s interval. only one metric is supported in the first version.
- metrics endpoint url: `LLMAutoScaler.Spec.Metrics[0].Source.Protocol://{pod ip}:LLMAutoScaler.Spec.Metrics[0].Source.Port/metrics`, default value is: `http://{pod ip}:5000/metrics`
- pod ip: get ip from pods that specified by InferenceAutoScaler.Spec.Metrics[0].Metric.Selector
- resolve metric value from response by `LLMAutoScaler.Spec.Metrics[0].Source.Name`, and default metric name is: `vllm:num_requests_waiting`
- If there are multiple pods are selected, average metric value should be calculated.
- If the specified metric can not resolved from the pod, for example, the pod is in the pending state. we should calculate the average value as following rules in order to prevent flapping.
  - In scale up direction: use the 0 as the metric value for missing pods.
  - In scale down direction: use the `LLMAutoScaler.Spec.Metrics[0].Threshold.High` as the metric value for missing pods.

### Basic Scaler

> The baisc scaler is used to scale GPU workloads accroding to specified metric changes. The scaling rules are shown as following:

Reviewer 4: ```suggestion
The basic scaler is used to scale GPU workloads according to specified metric changes. The scaling rules are shown as following:
```

Author: fixed

| item | scale up | scale down | introducation |
|----------|----------|----------|----------|
| scale step    | 1   | 1   |  only increase/reduce one replica in a scaling action, because the cost of gpu resource is really high |
| cooldown seconds    | 600   | 1800   | a waiting period after a scaling action for preventing frequent scaling |
| stablizationwindow seconds    | 30   | 300   | a lookback window that delays scaling decisions for avoiding premature downscaling. It means that scale-up is triggered only when the metric value exceeds the high threshold in 2 consecutive periods. and scale-down is triggered only when the metric value less than low threshold in 20 consecutive periods |

All these values for basic scale strategy is pre-configured according to the experience. so the values maybe adjusted if more suitable values are found in the future use.

The Scale Strategy Pseudocode

Inputs:

- CurrentReplicas: Actual number of replicas for target workload, resolved from /scale subresource API.
- CurrentWaitingRequests: current waiting requests in inference queue, resolved from pods by metric scraper.

> - MinReplicas: The max number of replicas for target object, related field: `LLMAutoScaler.Spec.MinReplicas`

Reviewer 4: ```suggestion
- MinReplicas: The min number of replicas for target object, related field: `LLMAutoScaler.Spec.MinReplicas`
```

- HighThreshold: expected high threshold of waiting requests, related field: `LLMAutoScaler.Spec.Metrics[0].Threshold.High`
- LowThreshold: expected low threshold of waiting requests, related field: `LLMAutoScaler.Spec.Metrics[0].Threshold.Low`
- ScaleUpStep: the scale step of scaling up action, default value is 1
- ScaleDownStep: the scale step of scaling down action, default value is 1
- UpStabilizationWindowSeconds: the stabilization window seconds of scaling up action, default value is 30
- DownStabilizationWindowSeconds: the stabilization window seconds of scaling down action, default value is 300
- UpCoolDownSeconds: the cool down seconds of scaling up action, default value is 600
- DownCoolDownSeconds: the cool down seconds of scaling down action, default value is 1800
- LastScaleTime: the timestamp for the last scaling action, related field: `LLMAutoScaler.Status.LastScaleTime`

Outputs:

- DesiredReplicas: Desired number of replicas for target workload. and the value will be used for scaling workload through /Scale subresource api.

```go
// 1. calculate the elapsed time for cooldown check

> coolddownElapsed := now.Sub(LastScaleTime)

Reviewer 4: ```suggestion
cooldownElapsed := now.Sub(LastScaleTime)
```

Author: fixed

by the way, scale subresource api for workspace CRD is addressed in this proposal: 

## Motivation

LLM inference service is a baisc and widly-used feature in Kaito, and Kaito community interest in auto scaler for inference workloads continues to intensify, related issues: [#306](https://github.com/kaito-project/kaito/issues/306), [#1104](https://github.com/kaito-project/kaito/issues/1104).

From the technical perspective, It's a good idea to provide auto-scaler capability, becasue the auto-scaler of inference workloads dynamically adjusts the number of inference instances based on request volume--scaling up during traffic spikes to improve inference speed, and scaling down during low demand to minimize GPU resource waste.

The auto-scaler solution for kaito should be shown as following:

![auto-scaler](../../static/img/workspace-scale-api.png)

and we will divide this auto-scaler feature into two parts as following:

> - Part one: support scale subresource api for workspace, so different auto-scaler solutions such as KEDA, HPA, etc. can be integrated with Kaito to mamage inference workloads dynamically. This part is addressed in another proposal: https://github.com/kaito-project/kaito/pull/1184.

Author: fixed

> - Part two: support a customized auto-sacler for kaito. The auto-scaler is designed for a minimalistic configuration experience, with most parameters pre-tuned for optimal performance. This allows users to easily get started without requiring specialized knowledge of LLM. This part will be addressed in this proposal. we name this auto-scaler as llm-auto-scaler. 

Reviewer 2: ```suggestion
- Part two: support a customized auto-scaler for kaito. The auto-scaler is designed for a minimalistic configuration experience, with most parameters pre-tuned for optimal performance. This allows users to easily get started without requiring specialized knowledge of LLM. This part will be addressed in this proposal. we name this auto-scaler as llm-auto-scaler. 
```

To ensure ease of use, The specialized auto-scaler is hosted in a independent repo(kaito-project/llm-auto-scaler). at the same time, the llm-auto-scaler component can work with kaito without depending on any third-party components. 

### Goals

- llm-auto-scaler is a specialized auto-scaler for scaling gpu workloads automatically, and can integrate with kaito to work.
- It is flexible to support mulitple scale strategies, and only one basic scale strategy(scaling workloads according to metrics change) is supported in the first version.

### Non-Goals/Future Work

- Support cron scale strategy(like cron job) for llm-auto-sacler in future version.

> - Only support to configure one metric for basic scale strategy, mutiple metrics will be supported in future version.

Reviewer 2: ```suggestion
- Only support to configure one metric for basic scale strategy, multiple metrics will be supported in future version.
```

- The time efficiency of the auto-scaler is not within the scope of this proposal, as it is influenced by mutliple external factors, including GPU node provisioning, LLM image pulling, etc.
- Only support scale vllm workload, and non-vllm is not covered.

## Proposal

### LLM-Auto-Scaler Architecture

The llm-auto-scaler component scrapes metrics from inference pod according to configurations in LLMAutoScaler CRD, and scaler controller calculate desired replicas by integrating scraped metrics and scale strategy,
then scale workspace replicas through /scale subresource API. The detailed auto-scaler architecture is shown in the following figure:

![auto-scaler](../../static/img/llm-auto-scaler.png)

- LLMAutoScaler CRD: is used as auto-scaler configuration(including scale strategy, target reference, etc.) for specified workspace resource.

> - Metrics Scraper: a module of llm-auto-scaler and used for scraping metrics from inference pods.

Reviewer 4: Just FYI, this is in the [k8s AI conformance proposal doc](https://docs.google.com/document/d/1hXoSdh9FEs13Yde8DivCYjjXyxa7j4J8erjZPEGWuzc/edit?usp=sharing):

> MUST: HorizontalPodAutoscaler must function correctly for pods utilizing accelerators. This includes the ability to scale these Pods based on custom metrics relevant to AI/ML workloads.

and 

> MUST: Enable easy integration for collecting common metrics from AI job frameworks (e.g., Ray Jobs) and inference servers (e.g., KServe, vLLM), such as job status, throughput, latency. Ensure standard compliance for common metrics (see [example](https://opentelemetry.io/docs/specs/semconv/gen-ai/gen-ai-metrics/)) to ensure consistency and interoperability

Since we are just starting, we should try to use the same metric names as described in https://opentelemetry.io/docs/specs/semconv/gen-ai/gen-ai-metrics if applicable.

From the technical perspective, it's a good idea to provide auto-scaler capability because the auto-scaler of inference workloads dynamically adjusts the number of inference instances based on request volume—scaling up during traffic spikes to improve inference speed, and scaling down during low demand to minimize GPU resource waste. Furthermore, for workloads with predictable, recurring traffic patterns, a time-based scaler can proactively adjust capacity, ensuring resources are ready before they are needed.

The auto-scaler solution for Kaito should be shown as follows:

![auto-scaler](../../static/img/workspace-scale-api.png)

We will divide this auto-scaler feature into two parts as follows:

- Part one: support scale subresource API for workspace, so different auto-scaler solutions such as KEDA, HPA, etc. can be integrated with Kaito to manage inference workloads dynamically. This part is addressed in another proposal: https://github.com/kaito-project/kaito/pull/1184.

> - Part two: support a customized auto-scaler for Kaito. The auto-scaler is designed for a minimalistic configuration experience, with most parameters pre-tuned for optimal performance. This allows users to easily get started without requiring specialized knowledge of LLM. This part will be addressed in this proposal. We name this auto-scaler as llm-auto-scaler, which will support both reactive (metric-based) and proactive (time-based) scaling.

Author: and I want to check what's your meaning about two types of scaling? Deployment and LWS?

> - Part two: support a customized auto-scaler for Kaito. The auto-scaler is designed for a minimalistic configuration experience, with most parameters pre-tuned for optimal performance. This allows users to easily get started without requiring specialized knowledge of LLM. This part will be addressed in this proposal. We name this auto-scaler as llm-auto-scaler, which will support both reactive (metric-based) and proactive (time-based) scaling.

Author: @wangyira I've posted a new proposal for supporting auto-scaler based on keda. please take a look.
https://github.com/kaito-project/kaito/pull/1237

Each cron policy defines a `schedule` in standard cron format and a `targetReplicas` count. The autoscaler will continuously monitor the clock and, when the current time matches a policy's schedule, it will adjust the workload's replica count to the specified target.

- Take the following policies as an example:

```yaml
policies:
  # The "scale-up" policy increases the number of replicas to 5 at 8:00 AM every day
  # to handle peak traffic during business hours.
  - name: "scale-up"
    schedule: "0 8 * * *" # Cron schedule for 8:00 AM daily
    targetReplicas: 5
  # The "scale-down" policy reduces the number of replicas to 1 at 1 minute past midnight (00:01)
  # every day to conserve resources during off-peak hours.
  - name: "scale-down"
    schedule: "1 0 * * *" # Cron schedule for 00:01 AM daily
    targetReplicas: 1

> ```

Reviewer 5: Is a scale-down policy enforced when a scale-up policy is created, what happens when the user only creates scale-up policy but not scale-down policy?

Author: There are no enforced policy settings in current proposal. If no scale-down policy is configured, the replicas will be equal to MaxReplicas after some scaling actions.

)

func calculateDesiredReplicasCron(INPUTS) OUTPUTS {
	now := time.Now()
	var matchingReplicas []int32

	// Find all policies that match the current time
	for _, policy := range CronPolicies {
		schedule, err := cron.ParseStandard(policy.Schedule)
		if err != nil {
			// Log error and skip policy
			continue
		}

		nextRun := schedule.Next(now.Add(-time.Minute)) // Check if the last run was within the last minute
		if now.Truncate(time.Minute).Equal(nextRun.Truncate(time.Minute)) {
			matchingReplicas = append(matchingReplicas, policy.TargetReplicas)
		}
	}

	// If no policy matches, no change in replicas
	if len(matchingReplicas) == 0 {
		return CurrentReplicas
	}

	// If multiple policies match, pick the one with the highest replica count
	// to ensure capacity for the highest-demand schedule.
	desired := matchingReplicas[0]
	for i := 1; i < len(matchingReplicas); i++ {
		if matchingReplicas[i] > desired {
			desired = matchingReplicas[i]
		}
	}

	// Clamp the desired replicas within the min/max boundaries
	if desired < MinReplicas {
		desired = MinReplicas
	}
	if desired > MaxReplicas {
		desired = MaxReplicas
	}

	return desired
}
```

## Alternatives

### Native HPA

> Native HPA + Prometheus + Prometheus Adapter solution can also be used for scaling inference workloads of Kaito.

Reviewer 5: Not sure if I follow if this is already supported, what are its benefits and drawbacks compared to the new llm scaler that's being proposed
